src/metric/rank_metric.cc

```diff
@@ -385,6 +378,33 @@ class EvalRankWithCache : public Metric {
   }
 
   double Evaluate(HostDeviceVector<float> const& preds, std::shared_ptr<DMatrix> p_fmat) override {
+    double result{0.0};
+    if (p_fmat->Info().IsVerticalFederated()) {
+      // TODO(rongou): better abstraction for this.
```

Comment:
Seems not too messy to de-duplicate the code with the one in the common header.

Reply:
OK I merged the aggregator branch to take care of this. Not sure about the name, if you can come up with something better I'd be happy to change it.